[docs-only] add storage architecture overview (#1920)
owncloud/ocis: docs/extensions/storage/_index.md

geekdocCollapseSection: true
---

## Abstract

This service provides an oCIS extension that wraps [reva](https://github.com/cs3org/reva/) and adds an opinionated configuration to it.

## Architecture Overview

> The below diagram shows the ocis services and the contained reva services within as dashed boxes. In general:

Reviewer: oCIS, happens more than once :)

1. A request comes in at the proxy and is authenticated using oidc.
2. It is forwarded to the ocis frontend which handles ocs and ocdav requests by talking to the reva gateway using the CS3 API.
3. The gateway acts as a facade to the actual CS3 services: storage providers, user providers, group providers and sharing providers.

{{< svg src="extensions/storage/static/overview.drawio.svg" >}}

The dashed lines in the diagram indicate requests that are made to authenticate requests or lookup the storage provider:

> 1. After authenticating a request the proxy may either use the CS3 `userprovider` or the accounts service to fetch the user information that will be minted into the `x-access-token`.

Reviewer: ```suggestion
1. After authenticating a request, the proxy may either use the CS3 `userprovider` or the accounts service to fetch the user information that will be minted into the `x-access-token`.
```

> 2. The gateway will verify the JWT signature of the `x-access-token` or try to authenticate the request itself, eg. using a public link token.

Reviewer: ```suggestion
2. The gateway will verify the JWT signature of the `x-access-token` or try to authenticate the request itself, e.g. using a public link token.
```

{{< hint warning >}}

> The bottom part is lighter because we will deprecate it in favor of using only the CS3 user and group providers after moving some account functionality into reva and glauth. The metadata storage is not registered in the reva gateway to prevent seperate user metadata necessary for running the service from data that is being served directly.

Reviewer: `to prevent seperate` sounds like an error to me, unclear what this sentence suggests :)

Reviewer: Also `eg.` again, perhaps best to search&replace with `e.g.` in the whole file

{{< /hint >}}

## Terminology

Communication is hard. And clear communication is even harder. You may encounter the following terms throughout the documentation, in the code or when talking to other developers. Just keep in mind that whenever you hear or read *storage*, that term needs to be clarified, because on its own it is too vague. PR welcome.

### Resources

> A *resource* is a logical concept. It can be have different [types](https://cs3org.github.io/cs3apis/#cs3.storage.provider.v1beta1.ResourceType)

Reviewer: ```suggestion
A *resource* is a logical concept. Ressources can be of [different types](https://cs3org.github.io/cs3apis/#cs3.storage.provider.v1beta1.ResourceType):
```

- an actual *file*
- a *container*, eg. a folder or bucket
- a *symlink*, or
- a *reference* which can point to a resource in another *storage provider*

### References

A [*reference*](https://cs3org.github.io/cs3apis/#cs3.storage.provider.v1beta1.Reference) is a logical concept. It identifies a *resource* and consists of either
- a *path* based reference, used to identify a *resource* in the *namespace* of a *storage provider*. It must start with a `/`.

> - an [*id* based reference](https://cs3org.github.io/cs3apis/#cs3.storage.provider.v1beta1.ResourceId), uniquely identifying a *resounce* in the *namespace* of a *storage provider*. It consists of a `storage provider id` and an `opaque id`. The `storage provider id` must NOT start with a `/`.

Reviewer: ```suggestion
- an [*id* based reference](https://cs3org.github.io/cs3apis/#cs3.storage.provider.v1beta1.ResourceId), uniquely identifying a *resource* in the *namespace* of a *storage provider*. It consists of a `storage provider id` and an `opaque id`. The `storage provider id` must NOT start with a `/`.
```

{{< hint info >}}
The `/` is important because currenty the static *storage registry* uses a map to look up which *storage provider* is responsible for the resource. Paths must be prefixed with `/` so there can be no collisions between paths and storage provider ids in the same map.
{{< /hint >}}

{{< hint warning >}}
**Proposed Change**

A *reference* is a logical concept. It identifies a *resource* and consists of
a *root* and a *path* relative to that *root*. 
A *root* is a `storage space id` and a `logical uuid` of a resource inside the *storage space*:
```
<storages_pace>:<logical_uuid>:<relative_path>
`------------root------------´ `----path-----´
```

> Both, *root* and *path*, are optional. The `storage space id` should be a uuid to prevent references from breaking when a *user* or *storage space* gets renamed. These are all valid references:

Reviewer: ```suggestion
*root* and *path* are both optional. The `storage space id` should be an uuid to prevent references from breaking when a *user* or *storage space* gets renamed. The following are all valid references:
```

Author: `be a uuid`, uuid starts with an umlaut and is pronounced in the same way.

> 

Author: I think what you have in mind is closer to the currently used types of references in CS3:
- an ID based reference, consisting of `storageid` and `opaqueid`
- a path based reference, consisting of a `path`

Most requests come with path based references in the global namespace. The global namespace could use aliases for storage spaces instead of ids. The gateway will look up which storage provider is responsible for the path in the storage registry, adjust the path based reference by ltrimming the mount path ond the storeg provider will receive a path that is relative to its root.

For Id based references the gateway will also look up which storage provider is responsible in the storage registry. This time, the storage registry extracts the storageid (which identifies the storage space, not the storage provider) and either finds it in a manually configured map of storageid->addr:port or in a list of cached storage ids (with a fallback to ping all existing storage providers if they know the opaqueid / id based reference to build the cache). In this case either the gateway or the storage registry may have to replace the opaqueid with the path by calling the getPithByID method on the storage provider.

> | name | description |

Reviewer: ```suggestion
| Name | Description |
```

|------|-------------|
|`::/users/alice/projects/foo` | follow this path in the  global namespace (path based reference)|
|`home-alice::projects/foo` | in the *storage space* `home-alice`, start at the *root*, follow `projects/foo` |

> |`home-alice:c3cf23bb-8f47-4719-a150-1d25a1f6fb56:foo` | in the *storage space* `home-alice`, start at the resource identified by the `logical uuid` `c3cf23bb-8f47-4719-a150-1d25a1f6fb56`, follow `projects/foo` |

Reviewer: Sure `projects/` is right here?

|`home-alice:56f7ceca-e7f8-4530-9a7a-fe4b7ec8089a:` | in the *storage space* `home-alice`, start at the resource identified by the `logical uuid` `56f7ceca-e7f8-4530-9a7a-fe4b7ec8089a`, no relativ path to follow |
|`ee1687e5-ac7f-426d-a6c0-03fed91d5f62:56f7ceca-e7f8-4530-9a7a-fe4b7ec8089a:` | in the *storage space* `ee1687e5-ac7f-426d-a6c0-03fed91d5f62`, start at the resource identified by the `logical uuid` `56f7ceca-e7f8-4530-9a7a-fe4b7ec8089a`, no relativ path to follow (id based reference) |

> They all reference the same *resource* (if the `logical uuid` of `projects` is 

Reviewer: ```suggestion
All those examples reference the same *resource* (if the `logical uuid` of `projects` is 
```

`c3cf23bb-8f47-4719-a150-1d25a1f6fb56` and the `logical uuid` of `foo` is `56f7ceca-e7f8-4530-9a7a-fe4b7ec8089a`).

> To move *storage spaces* between *storage providers*, the `logical uuid` of a file must be settable by the system. This also allows importing file metadata during a migration and restoring correct file ids when restoring a backup.

Reviewer: "settable" by which system? Does it need to be editable, too?

Plus "restoring from backup", perhaps?

Author: I reworded this. should be clearer now.

{{< /hint >}}

### Namespaces
A *namespace* is a set of paths with a common prefix. There is a global namespace when making requests against the CS3 *gateway* and a local namespace for every *storage provider*. Currently, oCIS mounts these *storage providers* out of the box:

| mountpoint | served namespace |
|------------|------------------|
| `/home`    | currently logged in users home |
| `/users`   | all users, used to access collaborative shares |
| `/public`  | publicly shared files, used to access public links |

{{< hint warning >}}

> We plan to serve all shares the current user has access to under `/shares`.

Reviewer: ```suggestion
We plan to serve all shares that the current user has access to under `/shares`.
```

{{< /hint >}}

### Storage Drivers

A *storage driver* implements access to a *storage system*:

> It maps the *path* and *id* based CS3 *references* to an appropriate *storage system* specific reference, eg:

Reviewer: ```suggestion
It maps the *path* and *id* based CS3 *references* to an appropriate *storage system* specific reference, e.g.:
```

- eos file ids
- posix inodes or paths
- deconstructed filesystem nodes

### Storage Provider

A *storage provider* manages *resources* identified by a *reference*
by accessing a *storage system* with a *storage driver*.

{{< svg src="extensions/storage/static/storageprovider.drawio.svg" >}}

{{< hint warning >}}
**Proposed Change**
A *storage provider* manages multiple *storage spaces*

> by accessing a *storage system* with a *storage driver*.

Author: reva already works like this. sending grant requests to the storage providers needs to be enabled in the first place. The challenge is that if you know the fileid the gateway / registry currently do not check if you have access to them ... which is why we always send the grants to the storage provider and implemented permission checks in the ocis driver.

{{< /hint >}}
{{< svg src="extensions/storage/static/storageprovider-spaces.drawio.svg" >}}
{{< hint warning >}}

> By making *storage providers* aware of *storage spaces* we can get rid of the current `enablehome` flag / hack in reva. Furthermore, provisioning a new *storage space* becomes a generic operation, regardless if a new user home or a new project space needs to be provisioned. 

Reviewer: ```suggestion
By making *storage providers* aware of *storage spaces* we can get rid of the current `enablehome` flag / hack in reva. Furthermore, provisioning a new *storage space* becomes a generic operation, regardless of the need of provisioning a new user home or a new project space.
```

{{< /hint >}}

### Storage Registries

A *storage registry* manages the global *namespace*:

> it is used by the *gateway*

Reviewer: ```suggestion
It is used by the *gateway*
```

to look up `address` and `port` of the *storage provider*
that should handle a *reference*.

{{< svg src="extensions/storage/static/storageregistry.drawio.svg" >}}

{{< hint warning >}}

> **Proposed Change**

Reviewer: Can't build the docs locally, but I guess it won't render the correct linebreak here since you're going for sort-of-a-heading. You may want to go for a h4/h5 (`#### Proposed Changes`) or 

```markdown
**Proposed Change:**

...xyu
```

A *storage registry* manages the *namespace* for a *user*:

> it is used by the *gateway*

Reviewer: ```suggestion
It is used by the *gateway*
```

to look up `address` and `port` of the *storage provider*
that is currently serving a *storage space*.
{{< /hint >}}
{{< svg src="extensions/storage/static/storageregistry-spaces.drawio.svg" >}}
{{< hint warning >}}

> By making *storage registries* aware of *storage spaces* we can query them for a listing of all *storage spaces* a user has access to. Including his home, received shares, project folders or group drives. See [Add draft of adr for spaces API ocis#1827](https://github.com/owncloud/ocis/pull/1827) for more info.

Reviewer: ```suggestion
By making *storage registries* aware of *storage spaces* we can query them for a listing of all *storage spaces* a user has access to. Including his home, received shares, project folders or group drives. See [a WIP PR for spaces in the oCIS repo (#1827)](https://github.com/owncloud/ocis/pull/1827) for more info.
```

{{< /hint >}}

### Storage Spaces
A *storage space* is a logical concept:

> it is a tree of *resources*

Reviewer: ```suggestion
It is a tree of *resources*
```

with a single *owner* (*user* or *group*), 
a *quota* and *permissions*, identified by a `storage space id`.

{{< svg src="extensions/storage/static/storagespace.drawio.svg" >}}

> Examples would be every user's home storage space, project storage spaces or group storage spaces. While they all serve different purposes and may or may not have workflows like anti virus scanning enabled, we need a way to identify and manage these subtrees in a generic way. By creating a dedicated concept for them this becomes easier and literally makes the codebase cleaner.

Reviewer 2: We also found in discussions that each share of a **directory** spans up a new storage space because it is also just a space a user has access to.

For a single file share that is different - that is just a resource and not a space (IIRC).

> Examples would be every users home storage space, project storage spaces or group storage spaces. While they all serve different purposes and may or may not have workflows like anti virus scanning enabled, we need a way to identify and manage these subtrees in a generic way. By creating a dedicated concept for them this becomes easier and literally makes the codebase cleaner.

Reviewer: ```suggestion
Examples would be every user's home storage space, project storage spaces or group storage spaces. While they all serve different purposes and may or may not have workflows like anti virus scanning enabled, we need a way to identify and manage these subtrees in a generic way. By creating a dedicated concept for them this becomes easier and literally makes the codebase cleaner.
```

> Finally, a logical `storage space id` is not tied to a specific storage provider. When the *storage driver* supports it we can import existing files including their fileid, which makes it possible to move storage spaces between storage spaces to implement storage classes eg. with or without archival, workflows, on SSDs or HDDs.

Reviewer: "to move storage spaces between storage spaces" => ?

```suggestion
Finally, a logical `storage space id` is not tied to a specific storage provider. If the *storage driver* supports it, we can import existing files including their *file id*, which makes it possible to move storage spaces between storage spaces to implement storage classes, e.g. with or without archival, workflows, on SSDs or HDDs.
```
